File: C++/Stack/Implementation/Array/Stack.cpp
//
// Created by vivek on 01-01-2023.
//

#include <iostream>
#include "Stack.h"

template<class T>
Stack<T>::Stack(unsigned int size):size(size) {         // Constructor
    array.resize(size);
    top = -1;
}

template<class T>
void Stack<T>::push(T x) {          // Push operation: insert element in a stack
    if (this->top == this->size - 1)            // Stack full condition
        std::cerr << "Stack overflow!\n";
    else {
        this->top++;
        this->array.at(top) = x;
    }
}

template<class T>
void Stack<T>::pop() {          // Pop operation: remove an element from a stack
    if (empty())
        std::cerr << "Stack underflow!\n";
    else {
        array.pop_back();
        this->top--;
    }
}

template<>
int Stack<int>::peek() {            // Get top element of stack when stack is of int type
    return empty() ? -1 : array.at(top);
}

template<>
char Stack<char>::peek() {            // Get top element of stack when stack is of char type
    return empty() ? ' ' : array.at(top);
}

template<>
double Stack<double>::peek() {            // Get top element of stack when stack is of double type
    return empty() ? -1 : array.at(top);
}

template<>
std::string Stack<std::string>::peek() {            // Get top element of stack when stack is of string type
    return empty() ? "" : array.at(top);
}

File: Java/Stack/src/stack/implementation/array/Stack.java
package stack.implementation.array;

import java.util.ArrayList;
import java.util.List;

public class Stack {
    private List<Integer> array;
    private final int size;
    private int top;

    public Stack(int size) {
        this.size = size;
        array = new ArrayList<>(size);
        top = -1;
    }

    private void push(int x) {
        if (this.top == this.size - 1)
            System.out.println("Stack overflow!");
        else {
            this.top++;
            array.add(top, x);
        }
    }

    private boolean empty() {
        return this.top == -1;
    }

    private void pop() {
        if (empty())
            System.out.println("Stack underflow!");
        else {
            array.remove(this.top);
            this.top--;
        }
    }

    private int peek() {
        return empty() ? -1 : array.get(array.size() - 1);
    }

    @Override
    public String toString() {
        StringBuilder output = new StringBuilder();
        for (int i = this.top; i >= 0; i--) {
            output.append(this.array.get(i));
            output.append("\n");
        }
        return output.toString();
    }

    public static void main(String[] args) {
        Stack stk = new Stack(10);
        stk.push(1);
        stk.push(2);
        stk.push(3);
        stk.pop();
        stk.pop();
        stk.pop();
        stk.pop();
        System.out.println(stk.peek());
    }
}
